### PR TITLE
Grt: use multi_array and vectors

### PR DESCRIPTION
**Pull request by luis201420**

### CODE REVIEW

File: src/grt/src/fastroute/src/route.cpp [hunk 1]
```diff
@@ -730,51 +693,47 @@ void FastRouteCore::newrouteZ(int netID, int threshold)
...
               if (tmp > 0)
                 cost_lr_[0] += tmp;
             }
             for (i = 1; i < segHeight; i++) {
               cost_lr_[i] = cost_lr_[i - 1];
-              grid = (y1 + i - 1) * x_grid_;
-              tmp = v_edges_[grid + x1].est_usage - v_capacity_lb_
-                    + v_edges_[grid + x1].red;
+              tmp = v_edges_[(y1 + i - 1)][x1].est_usage - v_capacity_lb_
+                    + v_edges_[(y1 + i - 1)][x1].red;
```
Comment (maliberty) [inline src/grt/src/fastroute/src/route.cpp:715]: () are not needed

File: src/grt/src/fastroute/src/route.cpp [hunk 2]
```diff
@@ -730,51 +693,47 @@ void FastRouteCore::newrouteZ(int netID, int threshold)
...
               if (tmp > 0)
                 cost_lr_[i] += tmp;
-              tmp = v_edges_[grid + x2].est_usage - v_capacity_lb_
-                    + v_edges_[grid + x2].red;
+              tmp = v_edges_[(y1 + i - 1)][x2].est_usage - v_capacity_lb_
+                    + v_edges_[(y1 + i - 1)][x2].red;
```
Comment (maliberty) [inline src/grt/src/fastroute/src/route.cpp:719]: no ()

File: src/grt/src/fastroute/src/route.cpp [hunk 3]
```diff
@@ -730,51 +693,47 @@ void FastRouteCore::newrouteZ(int netID, int threshold)
...
-              tmp = v_edges_[grid + x2].est_usage - v_capacity_lb_
-                    + v_edges_[grid + x2].red;
+              tmp = v_edges_[(y2 + i - 1)][x2].est_usage - v_capacity_lb_
+                    + v_edges_[(y2 + i - 1)][x2].red;
               if (tmp > 0)
                 cost_lr_[i] += tmp;
-              tmp = v_edges_[grid + x1].est_usage - v_capacity_lb_
-                    + v_edges_[grid + x1].red;
+              tmp = v_edges_[(y2 + i - 1)][x1].est_usage - v_capacity_lb_
+                    + v_edges_[(y2 + i - 1)][x1].red;
```
Comment (maliberty) [inline src/grt/src/fastroute/src/route.cpp:736]: no ()

File: src/grt/src/fastroute/src/utility.cpp [hunk 4]
```diff
@@ -1185,8 +1174,7 @@ void FastRouteCore::check2DEdgesUsage()
   // check vertical edges
   for (int i = 0; i < y_grid_ - 1; i++) {
     for (int j = 0; j < x_grid_; j++) {
-      int grid = i * x_grid_ + j;
-      if (v_edges_[grid].usage >= max_v_edge_usage) {
+      if (v_edges_[i][j].usage >= max_v_edge_usage) {
```
[the remark this answers is not in the document]
Comment (maliberty) [reply src/grt/src/fastroute/src/utility.cpp:1177]: I looks like they are so please add comments to their definitions explaining this (likewise with layer,y,x for 3d).

**Comment by maliberty** [comment]
It is a very nice improvement in readability.  Thanks!

File: src/grt/src/fastroute/src/maze3D.cpp [hunk 5]
```diff
@@ -1760,8 +1755,8 @@ void FastRouteCore::mazeRouteMSMDOrder3D(int expand,
   corr_edge_3D_.resize(boost::extents[0][0][0]);
   pr_3D_.resize(boost::extents[0][0][0]);
 
-  delete[] heap1_3D_;
-  delete[] heap2_3D_;
+  heap1_3D_.clear();
+  heap2_3D_.clear();
```
[the remark this answers is not in the document]
Comment (luis201420) [reply src/grt/src/fastroute/src/maze3D.cpp:1759]: Those heap are also used in the function setupHeap3D

[the remark this answers is not in the document]
Comment (eder-matheus) [reply src/grt/src/fastroute/src/maze3D.cpp:1759]: If they're only used in `setupHeap3D` and `mazeRouteMSMDOrder3D`, you can pass them to `setupHeap3D` as parameters.

File: src/grt/src/fastroute/include/FastRoute.h [hunk 6]
```diff
@@ -213,7 +213,10 @@ class FastRouteCore
                             int ripupTHlb,
                             int ripupTHub,
                             int layerOrientation);
-  void setupHeap3D(int netID,
+  void setupHeap3D(
+                   std::vector<int*> &heap1_3D_,
+                   std::vector<short*> &heap2_3D_,
```
Comment (eder-matheus) [inline src/grt/src/fastroute/include/FastRoute.h:218]: Fix format here

Comment (maliberty) [inline src/grt/src/fastroute/include/FastRoute.h:218]: Drop the trailing _ as these are not class members

**Pull request merged**